### PR TITLE
Feature/BDN-453 Regulations fix

### PR DESCRIPTION
**Pull request by klukwist**
https://appodeal.atlassian.net/browse/BDN-453

### CODE REVIEW

**Review by AntonZarytski — APPROVED** [review]

File: adapter/admob/src/main/java/org/bidon/admob/ext/RegulationExt.kt [hunk 1]
```diff
@@ -16,5 +16,5 @@ internal fun Regulation.asBundle() = Bundle().apply {
     this@asBundle.gdprConsentString?.let {
         putString("IABConsent_ConsentString", it)
     }
-    putBoolean("IABConsent_SubjectToGDPR", this@asBundle.gdprConsent)
+    putBoolean("IABConsent_SubjectToGDPR", this@asBundle.gdprApplies)
```
Comment (da2gl) [inline adapter/admob/src/main/java/org/bidon/admob/ext/RegulationExt.kt:19]: https://github.com/AppLovin/AppLovin-MAX-SDK-Android/blob/3692633b4a5b51f9e5529a75f2c7a4ea9e0a40bd/Google/src/main/java/com/applovin/mediation/adapters/GoogleMediationAdapter.java#L906

@klukwist по admob не могу найти документации. 
как ты сетишь эти данные? откуда взял?)

File: adapter/applovin/src/main/java/org/bidon/applovin/ApplovinAdapter.kt [hunk 2]
```diff
@@ -63,9 +63,14 @@ class ApplovinAdapter :
...
 
     override fun updateRegulation(regulation: Regulation) {
-        context?.let { context ->
-            AppLovinPrivacySettings.setHasUserConsent(regulation.gdprConsent, context)
-            AppLovinPrivacySettings.setIsAgeRestrictedUser(regulation.coppaApplies, context)
+        if (regulation.gdprApplies) {
+            AppLovinPrivacySettings.setHasUserConsent(regulation.hasGdprConsent, context)
+        }
+        if (regulation.ccpaApplies) {
+            AppLovinPrivacySettings.setDoNotSell(regulation.hasCcpaConsent, context)
```
Comment (da2gl) [inline adapter/applovin/src/main/java/org/bidon/applovin/ApplovinAdapter.kt]: !regulation.hasCcpaConsent should be inverted. 
https://dash.applovin.com/documentation/mediation/android/getting-started/privacy#multi-state-consumer-privacy-laws

File: adapter/bigoads/src/main/java/org/bidon/bigoads/BigoAdsAdapter.kt [hunk 3]
```diff
@@ -82,17 +81,13 @@ class BigoAdsAdapter :
...
         context?.let { context ->
-            BigoAdSdk.setUserConsent(
-                context,
-                consentOptions,
-                given
-            )
+            if (regulation.gdprApplies) {
+                BigoAdSdk.setUserConsent(context, ConsentOptions.GDPR, regulation.hasGdprConsent)
+            }
+            if (!regulation.usPrivacyString.isNullOrBlank()) {
```
Comment (da2gl) [inline adapter/bigoads/src/main/java/org/bidon/bigoads/BigoAdsAdapter.kt]: `if (regulation.ccpaApplies)` ?

File: adapter/dtexchange/src/main/java/org/bidon/dtexchange/DTExchangeAdapter.kt [hunk 4]
```diff
@@ -86,9 +86,15 @@ class DTExchangeAdapter :
...
 
     override fun updateRegulation(regulation: Regulation) {
-        InneractiveAdManager.setUSPrivacyString(regulation.usPrivacyString)
-        InneractiveAdManager.setGdprConsent(regulation.gdprConsent)
-        InneractiveAdManager.setGdprConsentString(regulation.gdprConsentString)
+        if (!regulation.usPrivacyString.isNullOrBlank()) {
+            InneractiveAdManager.setUSPrivacyString(regulation.usPrivacyString)
+        }
+        if (regulation.gdprApplies) {
+            InneractiveAdManager.setGdprConsent(true)
```
Comment (da2gl) [inline adapter/dtexchange/src/main/java/org/bidon/dtexchange/DTExchangeAdapter.kt]: `InneractiveAdManager.setGdprConsent(regulation.hasGdprConsent)` ?

Comment (da2gl) [reply adapter/dtexchange/src/main/java/org/bidon/dtexchange/DTExchangeAdapter.kt]: review`InneractiveAdManager.clearGdprConsentData()` / `InneractiveAdManager.clearUSPrivacyString()` / `InneractiveAdManager.currentAudienceAppliesToCoppa()` logic

File: adapter/dtexchange/src/main/java/org/bidon/dtexchange/DTExchangeAdapter.kt [hunk 5]
```diff
@@ -86,9 +86,15 @@ class DTExchangeAdapter :
     }
 
     override fun updateRegulation(regulation: Regulation) {
-        InneractiveAdManager.setUSPrivacyString(regulation.usPrivacyString)
-        InneractiveAdManager.setGdprConsent(regulation.gdprConsent)
-        InneractiveAdManager.setGdprConsentString(regulation.gdprConsentString)
+        if (!regulation.usPrivacyString.isNullOrBlank()) {
```
Comment (da2gl) [inline adapter/dtexchange/src/main/java/org/bidon/dtexchange/DTExchangeAdapter.kt]: `if (regulation.ccpaApplies)` ?

File: adapter/inmobi/src/main/java/org/bidon/inmobi/InmobiAdapter.kt [hunk 6]
```diff
@@ -111,6 +102,20 @@ class InmobiAdapter :
...
+
+    private fun getConsentObject(regulation: Regulation): JSONObject {
+        return JSONObject().apply {
+            if (regulation.gdpr != Gdpr.Unknown) {
+                this.put("gdpr", regulation.gdpr.code)
+            }
+            if (regulation.gdprApplies) {
+                this.put(InMobiSdk.IM_GDPR_CONSENT_AVAILABLE, regulation.hasGdprConsent)
+            }
+            regulation.iab.tcfV2?.let {
```
Comment (da2gl) [inline adapter/inmobi/src/main/java/org/bidon/inmobi/InmobiAdapter.kt]: `regulation.gdprConsentString` ?
may be now we don't need to obtain iab object?

File: adapter/meta/src/main/java/org/bidon/meta/MetaAudienceAdapter.kt [hunk 7]
```diff
@@ -76,7 +76,9 @@ class MetaAudienceAdapter :
     }
 
     override fun updateRegulation(regulation: Regulation) {
-        AdSettings.setMixedAudience(regulation.coppaApplies)
+        if (regulation.coppaApplies) {
+            AdSettings.setMixedAudience(true)
+        }
```
Comment (da2gl) [inline adapter/meta/src/main/java/org/bidon/meta/MetaAudienceAdapter.kt:81]: https://developers.facebook.com/docs/audience-network/optimization/best-practices/data-processing-options/

File: adapter/mobilefuse/src/main/java/org/bidon/mobilefuse/ext/GetMobileFuseTokenUseCase.kt [hunk 8]
```diff
@@ -19,7 +19,7 @@ internal object GetMobileFuseTokenUseCase {
         val tokenRequest = MobileFuseBiddingTokenRequest(
             privacyPreferences = MobileFusePrivacyPreferences.Builder()
                 .setSubjectToCoppa(regulation.coppaApplies)
-                .setSubjectToGdpr(regulation.gdprConsent)
+                .setSubjectToGdpr(regulation.gdprApplies)
```
Comment (da2gl) [inline adapter/mobilefuse/src/main/java/org/bidon/mobilefuse/ext/GetMobileFuseTokenUseCase.kt]: may be set like  in adapter/mobilefuse/src/main/java/org/bidon/mobilefuse/MobileFuseAdapter.kt?

File: adapter/vungle/src/main/java/org/bidon/vungle/VungleAdapter.kt [hunk 9]
```diff
@@ -73,15 +72,25 @@ class VungleAdapter :
...
+        if (regulation.gdprApplies) {
+            val status = if (regulation.hasGdprConsent) {
+                Vungle.Consent.OPTED_IN
+            } else {
+                Vungle.Consent.OPTED_OUT
+            }
+            Vungle.updateConsentStatus(status, null)
+        }
+        if (regulation.coppaApplies) {
+            Vungle.updateUserCoppaStatus(regulation.coppaApplies)
```
Comment (da2gl) [inline adapter/vungle/src/main/java/org/bidon/vungle/VungleAdapter.kt]: just true like another adapter
`Vungle.updateUserCoppaStatus(true)`

**Review by da2gl — CHANGES_REQUESTED** [review]

File: adapter/dtexchange/src/main/java/org/bidon/dtexchange/DTExchangeAdapter.kt (hunk 4 above)
Comment (klukwist) [reply adapter/dtexchange/src/main/java/org/bidon/dtexchange/DTExchangeAdapter.kt]: Added clearing

File: adapter/inmobi/src/main/java/org/bidon/inmobi/InmobiAdapter.kt (hunk 6 above)
Comment (klukwist) [reply adapter/inmobi/src/main/java/org/bidon/inmobi/InmobiAdapter.kt]: removed

File: adapter/mobilefuse/src/main/java/org/bidon/mobilefuse/ext/GetMobileFuseTokenUseCase.kt (hunk 8 above)
Comment (klukwist) [reply adapter/mobilefuse/src/main/java/org/bidon/mobilefuse/ext/GetMobileFuseTokenUseCase.kt]: refactored to 
`fun Regulation.toMobileFusePrivacyPreferences(): MobileFusePrivacyPreferences`

File: adapter/inmobi/src/main/java/org/bidon/inmobi/InmobiAdapter.kt (hunk 6 above)
Comment (da2gl) [reply adapter/inmobi/src/main/java/org/bidon/inmobi/InmobiAdapter.kt]: @klukwist I mean 
```
if (!regulation.gdprConsentString.isNullOrEmpty()) {
    this.put(InMobiSdk.IM_GDPR_CONSENT_IAB, regulation.gdprConsentString)
}
```
instead 
`regulation.iab.tcfV2`

File: adapter/meta/src/main/java/org/bidon/meta/MetaAudienceAdapter.kt (hunk 7 above)
Comment (da2gl) [reply adapter/meta/src/main/java/org/bidon/meta/MetaAudienceAdapter.kt:81]: @klukwist 
by docs it should have empty array 
just make it like [Bidmachine Adapter](https://github.com/bidmachine/BidMachine-Android-SDK/blob/cf4febf4e14aa338a18f4f8a485f92a406d5b7f6/adapters/meta_audience/src/main/java/io/bidmachine/ads/networks/meta_audience/MetaAudienceAdapter.java#L125)

File: adapter/inmobi/src/main/java/org/bidon/inmobi/InmobiAdapter.kt (hunk 6 above)
Comment (klukwist) [reply adapter/inmobi/src/main/java/org/bidon/inmobi/InmobiAdapter.kt]: I'm not sure we should put publisher value for this `IM_GDPR_CONSENT_IAB`, shouldn't?

Comment (klukwist) [reply adapter/inmobi/src/main/java/org/bidon/inmobi/InmobiAdapter.kt]: Added

Comment (da2gl) [reply adapter/inmobi/src/main/java/org/bidon/inmobi/InmobiAdapter.kt]: if anything happens, put the blame on me

**Review by da2gl — APPROVED** [review]

**Pull request merged**